### ISSUE TITLE
Inconsistent behavior for empty records

### ISSUE THREAD

**Issue by merlea** [MDSplus/mdsplus#2401]
I am experiencing some issues using the latest stable version, when using empty arrays (actually of shape `[0,0]`).

I first create the tree with the following TCL script
```
edit test/new
add node data/usage=numeric
write
quit
```

I then test putting some data and retrieving both locally and with mdsip, first with `tdic`:
```
TDI> TreeOpen("test",-1);
TDI> %
TDI> mdsconnect('localhost');
TDI> mdsvalue('setenv("test_path=/home/amerle")');
TDI> mdsopen("test",-1);
TDI> %
TDI> TreePut(":DATA","$1",SET_RANGE(0,0,0));
TDI> Tcl("show data :data");
\TEST::TOP:DATA
 DTYPE_L                       :  Array [ 0,0 ]
TDI> write(*,"local");
local
TDI> shape(:DATA)
[0,0]
TDI> shape(getnci(":DATA","record"))
%TDI Error in GETNCI(":DATA", "record")
%TDI Error in SHAPE(GETNCI(":DATA", "record"))
%TDI Error in EXECUTE('shape(getnci(":DATA","record"))')

TDI> write(*,"remote return value");
remote return value
TDI> shape(mdsvalue(':DATA'))
[]
TDI> shape(mdsvalue('getnci(":DATA","record")'))
[]
TDI> write(*,"remote value");
remote value
TDI> mdsvalue('shape(:DATA)')
[0,0]
TDI> mdsvalue('shape(getnci(":DATA","record"))')
*
TDI> %
TDI> TreePut(":DATA","SET_RANGE(0,0,0)");
TDI> Tcl("show data :data");
\TEST::TOP:DATA
 DTYPE_FUNCTION                : OPC$SET_RANGE
     DTYPE_L                       : 0 
     DTYPE_L                       : 0 
     DTYPE_L                       : 0 
TDI> write(*,"local");
local
TDI> shape(:DATA)
[0,0]
TDI> shape(getnci(":DATA","record"))
[0,0]
TDI> write(*,"remote return value");
remote return value
TDI> shape(mdsvalue(':DATA'))
[]
TDI> shape(mdsvalue('getnci(":DATA","record")'))
[]
TDI> write(*,"remote value");
remote value
TDI> mdsvalue('shape(:DATA)')
[0,0]
TDI> mdsvalue('shape(getnci(":DATA","record"))')
[0,0]
TDI> %
TDI> TreePut(":DATA","$1","ARRAY([0,0])");
TDI> Tcl("show data :data");
\TEST::TOP:DATA
 DTYPE_FUNCTION                : OPC$SET_RANGE
     DTYPE_L                       : 0 
     DTYPE_L                       : 0 
     DTYPE_L                       : 0 
TDI> write(*,"local");
local
TDI> shape(:DATA)
[0,0]
TDI> shape(getnci(":DATA","record"))
[0,0]
TDI> write(*,"remote return value");
remote return value
TDI> shape(mdsvalue(':DATA'))
[]
TDI> shape(mdsvalue('getnci(":DATA","record")'))
[]
TDI> write(*,"remote value");
remote value
TDI> mdsvalue('shape(:DATA)')
[0,0]
TDI> mdsvalue('shape(getnci(":DATA","record"))')
[0,0]
TDI> %
TDI> TreePut(":DATA","ARRAY([0,0])");
TDI> Tcl("show data :data");
\TEST::TOP:DATA
 DTYPE_FUNCTION                : OPC$ARRAY
     DTYPE_L                       :  Array [ 2 ]
TDI> write(*,"local");
local
TDI> shape(:DATA)
[0,0]
TDI> shape(getnci(":DATA","record"))
[0,0]
TDI> write(*,"remote return value");
remote return value
TDI> shape(mdsvalue(':DATA'))
[]
TDI> shape(mdsvalue('getnci(":DATA","record")'))
[]
TDI> write(*,"remote value");
remote value
TDI> mdsvalue('shape(:DATA)')
[0,0]
TDI> mdsvalue('shape(getnci(":DATA","record"))')
[0,0]
```
So first accessing the tree locally (thick client) there is sometimes an error using `getnci(":DATA","record")`. Then, reading it remotely it seems the return value is always `$Missing` and thus of shape `[]` instead of `[0,0]`.

I also tested this using the Python interface with the following script:
```
#/usr/bin/env python

import MDSplus

tree = MDSplus.Tree("test",-1);
node = tree.getNode(":DATA");

conn = MDSplus.Connection('localhost');
conn.get('setenv("test_path=/home/amerle")');
conn.openTree("test",-1);

for ii in range(0,4):
    if ii == 0:
        node.putData(node.execute("SET_RANGE(0,0,0)"));
    elif ii == 1:
        node.putData(node.compile("SET_RANGE(0,0,0)"));
    elif ii == 2:
        node.putData(node.execute("ARRAY([0,0])"));
    elif ii == 3:
        node.putData(node.compile("ARRAY([0,0])"));

    tree.tcl("show data :data");
    print("local");
    try:
        print(node.getShape())
    except:
        print("Exception while retrieving node shape")
    try:
        print(node.getRecord().getShape())
    except:
        print("Exception while retrieving record shape")

    print("remote return value");
    try:
        print(conn.get(':DATA').getShape())
    except:
        print("Exception while retrieving node shape")
    try:
        print(conn.get('getnci(":DATA","record")').getShape())
    except:
        print("Exception while retrieving record shape")
    print("remote value");
    try:
        print(conn.get('shape(:DATA)'))
    except:
        print("Exception while retrieving node shape")
    try:
        print(conn.get('shape(getnci(":DATA","record"))'))
    except:
        print("Exception while retrieving record shape")
```
with the following result:
```
\TEST::TOP:DATA
 DTYPE_L                       :  Array [ 0,0 ]

local
[0,0]
[0,0]
remote return value
Exception while retrieving node shape
Exception while retrieving record shape
remote value
[0,0]
Exception while retrieving record shape
\TEST::TOP:DATA
 DTYPE_FUNCTION                : OPC$SET_RANGE
     DTYPE_L                       : 0 
     DTYPE_L                       : 0 
     DTYPE_L                       : 0 

local
[0,0]
[0,0]
remote return value
Exception while retrieving node shape
Exception while retrieving record shape
remote value
[0,0]
[0,0]
\TEST::TOP:DATA
 DTYPE_FS                      :  Array [ 0,0 ]

local
[0,0]
[0,0]
remote return value
Exception while retrieving node shape
Exception while retrieving record shape
remote value
[0,0]
Exception while retrieving record shape
\TEST::TOP:DATA
 DTYPE_FUNCTION                : OPC$ARRAY
     DTYPE_L                       :  Array [ 2 ]

local
[0,0]
[0,0]
remote return value
Exception while retrieving node shape
Exception while retrieving record shape
remote value
[0,0]
[0,0]
```
Here accessing the data locally is always successful but accessing them through MDSip yields some errors.

Do you have any idea what might be happening?

**Comment by merlea** [comment]
I have been looking around with debugging enabled and found 2 issues when dealing with empty records that explain the above behaviour, one in `MdsIpShr`, the other in `TdiShr`. They are described below along with the 2 solutions I have found. With these implemented the tests above behave as expected.

Please let me know what you think of the solutions as my MDSplus inner workings is quite limited and had to guess a lot. I can also open a Pull Request if that's easier.

### Issue in `MdsIpShr`
In `GetAnswerInfoTO`, it seems that if the `dims` field of the received message has zeros in it, then some kind of recovery is attempted to guess what the `dims` should have been which failed.This resulted in a mis-calculation of `numbytes` with a result different than `datalen` and so an error was triggered.  But here the zeros are meant to be there as `datalen` is also zero. The following patch allowed the routine to complete without errors for various kinds of empty values:
```diff
diff --git a/mdstcpip/mdsipshr/GetAnswerInfo.c b/mdstcpip/mdsipshr/GetAnswerInfo.c
index e77f407cc..e035dbc38 100644
--- a/mdstcpip/mdsipshr/GetAnswerInfo.c
+++ b/mdstcpip/mdsipshr/GetAnswerInfo.c
@@ -88,21 +88,21 @@ int GetAnswerInfoTO(int id, char *dtype, short *length, char *ndims, int *dims,
   int datalen = (int)(m->h.msglen - sizeof(MsgHdr));
   *numbytes = m->h.length;
   if (m->h.ndims)
   {
     for (i = 0; i < m->h.ndims; i++)
     {
 #ifdef __CRAY
       dims[i] =
           i % 2 ? m->h.dims[i / 2] & 0xffffffff : (*m)->h.dims[i / 2] >> 32;
 #else
-      if (m->h.dims[i])
+      if (m->h.dims[i] || !datalen)
         dims[i] = m->h.dims[i];
       else
       {
         if (datalen > *numbytes && datalen % *numbytes)
         {
           dims[i] = datalen / *numbytes;
           m->h.ndims = i + 1;
         }
         else if (datalen == *numbytes && m->h.ndims == i + 1)
         {
```
(In the example above, we would have `h.length=4,datalen=0, h.ndims=2, dims=[0,0,0,0,0,0,0,0],...`). This patch allows dimensions equal to 0 if the data portion of the message is of size 0.

### Issue in TdiShr
Here the problem lies in `Tdi1GetNci`. When doing `getnci("..:","record")` on a node with a record containing an empty value, the very last check in the function was causing a `TreeNNF` (Node-Not-Found?) error. If I understand correctly this last check is for cases where no error has been encountered, a record of type array has been found but its size is 0. I am only guessing here, but maybe this was meant to catch cases when no data was actually stored for this node (and thus the record length would be 0 too). 
So here I added the condition that the record length had to be zero too to raise an error. I also changed the error to be a `TreeNODATA` error as this seemed more appropriate than `TreeNNF`. Here is the patch:
```diff
diff --git a/tdishr/TdiGetNci.c b/tdishr/TdiGetNci.c
index 832532851..ff6735a7b 100644
--- a/tdishr/TdiGetNci.c
+++ b/tdishr/TdiGetNci.c
@@ -648,15 +648,16 @@ int Tdi1GetNci(opcode_t opcode __attribute__((unused)), int narg,
       MdsFree1Dx(&tmp, NULL);
     }
   }
   if (outcount && key_ptr->item_length == 0)
     for (j = outcount, hold_ptr -= step; --j >= 0; hold_ptr -= step)
       MdsFree1Dx((struct descriptor_xd *)hold_ptr, NULL);
   MdsFree1Dx(&holdxd, NULL);
   if (STATUS_OK && key_ptr->item_dtype == DTYPE_NID && out_ptr->pointer)
     out_ptr->pointer->dtype = DTYPE_NID;
   if (STATUS_OK && out_ptr->pointer && out_ptr->pointer->class == CLASS_A &&
+      out_ptr->pointer->length == 0 &&
       ((struct descriptor_a *)(out_ptr->pointer))->arsize == 0)
-    status = TreeNNF;
+    status = TreeNODATA;
 
   return status;
 }
```
In the example above `out_ptr->pointer->length=44`.

**Comment by merlea** [comment]
I opened the pull request #2408 to discuss this.

**Comment by merlea** [comment]
Closed by #2408
